qubes-staging molecule scenario fails at create step

## Description

qubes-staging molecule action fails to create (clone sd-$ROLE-base) in the adminvm:
```
"stderr": "app: Failed to clone appmenus, qvm-appmenus missing\nqvm-clone: error: Failed to clone appmenus", "stderr_lines": ["app: Failed to clone appmenus, qvm-appmenus missing", "qvm-clone: error: Failed to clone appmenus"], "stdout": "", "stdout_lines": []}
```
manually running qvm-clone results in the same error (and the VMs are not cloned):
```
app: Failed to clone appmenus, qvm-appmenus missing
qvm-clone: error: Failed to clone appmenus
```
## Steps to Reproduce

1. Install staging environment in qubes (using https://docs.securedrop.org/en/release-0.10.0/development/qubes_staging.html)
2. `molecule converge -s qubes-staging`
3. observe error described above
## Expected Behavior

molecule scenario should succeed and a SecureDrop staging environment should be created
## Actual Behavior

Molecule scenario fails at the Create step

## Comments

qvm-clone automatically uses the class of the source of the copy (`--class StandaloneVM`), and passing `--class StandaloneVM` does not address the issue.

Workaround is to qvm-clone the base VMs in dom0, and then run the molecule converge step in the adminVM. In dom0:
```
qvm-clone sd-app-base sd-app
qvm-tags sd-app add created-by-sd-dev
```

@rmol I'm definitely still seeing the error locally:

<blockquote>
  TASK [Clone base image for staging VMs] ****************************************
    fatal: [localhost]: FAILED! => {"msg": "The conditional check 'clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines' failed. The error was: error while evaluating conditional (clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines): Unable to look up a name or access an attribute in template string ({% if clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines %} True {% else %} False {% endif %}).\nMake sure your variable name does not contain invalid characters like '-': argument of type 'StrictUndefined' is not iterable"}
    	to retry, use: --limit @/home/user/securedrop/molecule/qubes-staging/create.retry
</blockquote>

For complete's sake, here's confirmation that I'm running latest develop, with latest pip deps installed.


<details>

<summary>Terminal session</summary>

```
[user@sd-dev:~/securedrop] [sd] develop+ 39s ± 
 $ git status -sb
## develop...origin/develop
[user@sd-dev:~/securedrop] [sd] develop+ 43s ± 
 $ git rev-parse HEAD
a0e26746ade9e52a45785b167859f2d3e6c654a0
[user@sd-dev:~/securedrop] [sd] develop+ 45s ± 
 $ pip install -r securedrop/requirements/python3/develop-requirements.txt  > /dev/null
WARNING: You are using pip version 19.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
[user@sd-dev:~/securedrop] [sd] develop+ 50s ± 
 $ make staging
 Creating staging environment...
Creating staging environment via 'qubes-staging'...
Using active Python 3 virtualenv in /home/user/.virtualenvs/sd
--> Validating schema /home/user/securedrop/molecule/fetch-tor-packages/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/libvirt-staging-xenial/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/ansible-config/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/builder-xenial/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/vagrant-packager/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/virtualbox-staging-xenial/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/qubes-staging/molecule.yml.
Validation completed successfully.
--> Validating schema /home/user/securedrop/molecule/upgrade/molecule.yml.
Validation completed successfully.
--> Test matrix
    
└── qubes-staging
    ├── dependency
    ├── create
    ├── prepare
    └── converge
    
--> Scenario: 'qubes-staging'
--> Action: 'dependency'
Skipping, missing the requirements file.
--> Scenario: 'qubes-staging'
--> Action: 'create'
    
    PLAY [Create] ******************************************************************
    
    TASK [Gathering Facts] *********************************************************
    ok: [localhost]
    
    TASK [Clone base image for staging VMs] ****************************************
    fatal: [localhost]: FAILED! => {"msg": "The conditional check 'clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines' failed. The error was: error while evaluating conditional (clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines): Unable to look up a name or access an attribute in template string ({% if clone_result.rc != 0 and \"qvm-clone: error: VM \"+item.vm_name+\" already exists\" not in clone_result.stderr_lines %} True {% else %} False {% endif %}).\nMake sure your variable name does not contain invalid characters like '-': argument of type 'StrictUndefined' is not iterable"}
    	to retry, use: --limit @/home/user/securedrop/molecule/qubes-staging/create.retry
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=1    changed=0    unreachable=0    failed=1   
    
ERROR: 
make: *** [Makefile:200: staging] Error 2
```

I use a patch to get around this, so I'll submit that, as well as knocking out the longstanding #3629.

--- a/docs/development/qubes_staging.rst
+++ b/docs/development/qubes_staging.rst
@@ -299,12 +299,6 @@ environment. In from the root of the SecureDrop project in ``sd-dev``, run:
 
    make staging
 
-One limitation of Qubes is that the reboot handlers which run during 
-provisioning can shut down the VMs, but not start them again. When you see the
-message ``RUNNING HANDLER [common : Wait for server to come back.]`` you must 
-start the VMs again manually from ``dom0`` with the command 
-``qvm-start sd-staging-app && qvm-start sd-staging-mon``.
-
 The ``make staging`` command invokes the ``qubes-staging`` Molecule scenario. 
 You can also run constituent Molecule actions directly, rather than using
 the Makefile target: 

--- a/install_files/ansible-base/tasks/reboot.yml
+++ b/install_files/ansible-base/tasks/reboot.yml
@@ -3,3 +3,23 @@
   reboot:
   tags:
     - reboot
+  when: not securedrop_staging_qubes_env|default(False)
+
+- name: Gracefully halt Qubes staging VM
+  command: qvm-shutdown --wait {{ "sd-staging-app" if "app" in inventory_hostname else "sd-staging-mon" }}
+  become: no
+  delegate_to: localhost
+  when: securedrop_staging_qubes_env|default(False)
+  tags:
+    - reboot
+
+- name: Boot Qubes staging VM
+  shell: >
+    qvm-start
+    {{ "sd-staging-app" if "app" in inventory_hostname else "sd-staging-mon" }}
+    && sleep 30
+  become: no
+  delegate_to: localhost
+  when: securedrop_staging_qubes_env|default(False)
+  tags:
+    - reboot

--- a/molecule/qubes-staging/create.yml
+++ b/molecule/qubes-staging/create.yml
@@ -9,7 +9,9 @@
   tasks:
 
     - name: Clone base image for staging VMs
-      command: qvm-clone {{ item.vm_base }} {{ item.vm_name }}
+      # The "ignore-errors" flag sidesteps an issue with qvm-sync-appmenus. We don't need
+      # app menus for the SD VMs, so an error there need not block provisioning.
+      command: qvm-clone {{ item.vm_base }} {{ item.vm_name }} --ignore-errors
       register: clone_result
       failed_when: >-
         clone_result.rc != 0 and "qvm-clone: error: VM "+item.vm_name+" already exists" not in clone_result.stderr_lines

--- a/molecule/qubes-staging/create.yml
+++ b/molecule/qubes-staging/create.yml
@@ -7,6 +7,12 @@
     molecule_instance_config: "{{ lookup('env', 'MOLECULE_INSTANCE_CONFIG') }}"
     molecule_yml: "{{ lookup('file', molecule_file) | molecule_from_yaml }}"
   tasks:
+    - name: Check that Qubes admin tools are installed
+      shell: >
+        which qvm-clone
+        || { echo 'qvm-clone not found, install qubes-core-admin-client';
+        exit 1; }
+      changed_when: false
 
     - name: Clone base image for staging VMs
       # The "ignore-errors" flag sidesteps an issue with qvm-sync-appmenus. We don't need

--- a/molecule/qubes-staging/destroy.yml
+++ b/molecule/qubes-staging/destroy.yml
@@ -9,6 +9,13 @@
     molecule_yml: "{{ lookup('file', molecule_file) | molecule_from_yaml }}"
     molecule_ephemeral_directory: "{{ lookup('env', 'MOLECULE_EPHEMERAL_DIRECTORY') }}"
   tasks:
+    - name: Check that Qubes admin tools are installed
+      shell: >
+        which qvm-clone
+        || { echo 'qvm-clone not found, install qubes-core-admin-client';
+        exit 1; }
+      changed_when: false
+
     - name: Halt molecule instance(s)
       command: qvm-shutdown --wait "{{ item.vm_name }}"
       register: server

--- a/molecule/qubes-staging/qubes-vars.yml
+++ b/molecule/qubes-staging/qubes-vars.yml
@@ -23,3 +23,7 @@ remote_host_ref: >-
   }}
 
 securedrop_staging_install_target_distro: xenial
+
+# Inform the Ansible logic we're targeting Qubes staging VMs,
+# helps to customize the reboot logic.
+securedrop_staging_qubes_env: True

